pkg/sr/serde.go

```diff
@@ -316,44 +312,103 @@ func (s *Serde) DecodeNew(b []byte) (any, error) {
+	if err != nil {
+		return nil, tserde{}, err
 	}
-	id := binary.BigEndian.Uint32(b[1:5])
-	b = b[5:]
 
-	t := s.loadIDs()[int(id)]
+	t := s.loadIDs()[id]
 	if len(t.subindex) > 0 {
-		r := bReader{b}
-		br := io.ByteReader(&r)
-		l, err := binary.ReadVarint(br)
-		if l == 0 { // length 0 is a shortcut for length 1, index 0
-			t = t.subindex[0]
-		}
-		for err == nil && t.subindex != nil && l > 0 {
-			var idx int64
-			idx, err = binary.ReadVarint(br)
-			t = t.subindex[int(idx)]
-			l--
-		}
+		var index []int
+		index, b, err = h.DecodeIndex(b)
 		if err != nil {
-			return nil, t, err
+			return nil, tserde{}, err
+		}
+		for _, idx := range index {
+			if t.subindex == nil {
+				return nil, tserde{}, ErrNotRegistered
+			}
+			t = t.subindex[idx]
 		}
-		b = r.b
 	}
 	if !t.exists {
-		return nil, t, ErrNotRegistered
+		return nil, tserde{}, ErrNotRegistered
 	}
 	return b, t, nil
 }
 
-type bReader struct{ b []byte }
+// SerdeHeader encodes and decodes a message header.
+type SerdeHeader interface {
+	AppendEncode(b []byte, id int, index []int) ([]byte, error)
+	DecodeID(in []byte) (id int, out []byte, err error)
+	DecodeIndex(in []byte) (index []int, out []byte, err error)
+}
 
-func (b *bReader) ReadByte() (byte, error) {
-	if len(b.b) > 0 {
-		r := b.b[0]
-		b.b = b.b[1:]
-		return r, nil
+// ConfluentHeader is a SerdeHeader that produces the Confluent wire format. It
+// starts with 0, then big endian uint32 of the ID, then index (only protobuf),
+// then the encoded message.
+//
+// https://docs.confluent.io/platform/current/schema-registry/serdes-develop/index.html#wire-format
+type ConfluentHeader struct{}
+
+// AppendEncode appends an encoded header to b according to the Confluent wire
+// format and returns it. Error is always nil.
+func (ConfluentHeader) AppendEncode(b []byte, id int, index []int) ([]byte, error) {
+	b = append(
+		b,
+		0,
+		byte(id>>24),
+		byte(id>>16),
+		byte(id>>8),
+		byte(id>>0),
+	)
+
+	if len(index) > 0 {
+		if len(index) == 1 && index[0] == 0 {
+			b = append(b, 0) // first-index shortcut (one type in the protobuf)
+		} else {
+			b = binary.AppendVarint(b, int64(len(index)))
+			for _, idx := range index {
+				b = binary.AppendVarint(b, int64(idx))
+			}
+		}
+	}
+
+	return b, nil
+}
+
+// DecodeID strips and decodes the schema ID from b. It returns the ID alongside
+// the unread bytes. If the header does not contain the magic byte or b contains
+// less than 5 bytes it returns ErrBadHeader.
+func (ConfluentHeader) DecodeID(b []byte) (int, []byte, error) {
+	if len(b) < 5 || b[0] != 0 {
+		return 0, nil, ErrBadHeader
+	}
+	id := binary.BigEndian.Uint32(b[1:5])
+	return int(id), b[5:], nil
+}
+
+// DecodeIndex strips and decodes the index from b. It returns the index
+// alongside the unread bytes. It expects b to be the output of DecodeID (schema
+// ID should already be stripped away).
+func (ConfluentHeader) DecodeIndex(b []byte) ([]int, []byte, error) {
+	buf := bytes.NewBuffer(b)
```

Comment:
bytes.Reader

Reply:
I wanted to use the reader initially, but it lacks the `Bytes` method that the buffer provides, for returning unread bytes.

Reply:
wdyt about re-adding bReader in light of this^^?

Reply:
I can do that if you feel strongly about it 😉 I'm still curious why it's worth avoiding `bytes.Buffer`? I created a [simple benchmark](https://go.dev/play/p/Giaik-DzloL) for `bReader` and `bytes.Buffer`, both seem to have virtually the same performance and no allocations.
```
go test -bench=. -benchmem .
goos: darwin
goarch: amd64
pkg: github.com/twmb/franz-go/pkg/sr
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkBReader-16        	148830937	         7.766 ns/op	       0 B/op	       0 allocs/op
BenchmarkBytesBuffer-16    	149000121	         7.886 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/twmb/franz-go/pkg/sr	4.065s
```

Reply:
I don't feel too strongly since the benchmarks show things are the same